packages/core/src/MachineNode.ts

```diff
@@ -0,0 +1,351 @@
+import { toArray, isBuiltInEvent, toSCXMLEvent } from './utils';
+import {
+  Event,
+  StateValue,
+  MachineOptions,
+  EventObject,
+  TransitionDefinition,
+  DelayedTransitionDefinition,
+  StateSchema,
+  ActionObject,
+  Mapper,
+  PropertyMapper,
+  NullEvent,
+  MachineConfig,
+  SCXML,
+  Typestate,
+  TransitionDefinitionMap
+} from './types';
+import { State } from './State';
+
+import { toActionObject } from './actions';
+import { IS_PRODUCTION } from './environment';
+import { STATE_DELIMITER } from './constants';
+import { getConfiguration, getChildren, getAllStateNodes } from './stateUtils';
+import {
+  getStateNodeById,
+  getInitialState,
+  getStateNodes,
+  transitionNode,
+  resolveStateValue,
+  resolveTransition
+} from './nodeUtils';
+import { StateNode } from './StateNode';
+
+export const NULL_EVENT = '';
+export const STATE_IDENTIFIER = '#';
+export const WILDCARD = '*';
+
+const createDefaultOptions = <TContext>(
+  context: TContext
+): MachineOptions<TContext, any> => ({
+  actions: {},
+  guards: {},
+  services: {},
+  activities: {},
+  delays: {},
+  context
+});
+
+function resolveContext<TContext>(
+  context: TContext,
+  partialContext: Partial<TContext>
+): TContext {
+  if (context === undefined) {
```

Comment:
I think this might be a bug - don't we allow right now **not** defining context upfront (during "first" `Machine` call) and supplying it later for example through `withContext` or with second argument to `useMachine`? If yes - then this logic won't use the supplied context if the machine "template" has no context defined on it.

Reply:
This is something we have to think through, but we should _not_ allow context to be `undefined` whenever context is expected to be defined. This is because it's very possible to do `const someMachine = createMachine(...)` without ever defining context, and this is a footgun.

The current pattern for this (that I will encourage in the docs, unless we design another API around it) is to provide factories for machines with dynamic initial contexts:

```js
const createUserMachine = (userId) => createMachine({
  initial: 'pending',
  context: {
    userId
  },
  // ...
});
```

Reply:
This code is for the specific case of when the `context` should be `undefined` (context-less machine - only finite states)

Reply:
> This is something we have to think through, but we should not allow context to be undefined whenever context is expected to be defined.

Agreed, I've tried to address this in the past with https://github.com/davidkpiano/xstate/pull/588 but it was a breaking change back then so we couldn't pursue this further.

It should be fairly simple to introduce this constraint on the type-level and we can even introduce some runtime dev-only validations. I've added https://github.com/davidkpiano/xstate/issues/583 to v5.0 milestone so we don't miss this.